Spring Cloud Gateway: onInbound autoRead no keep-alive RST Error

<!--- Provide a general summary of the issue in the Title above -->

Jmeter <--> gateway <--> microservice (have a endpoints sleep 10s)

The request is sent to the gateway, which forwards the request to the microservice, which sleeps for 10 seconds. Jmeter then forces the request to end and send RST. I'm getting  **java.io.IOException: Connection reset by peer**


<!--- /!\ Make sure to follow the Contribution Guidelines and notably for security issues:
https://github.com/reactor/reactor-netty/blob/master/CONTRIBUTING.md
https://tanzu.vmware.com/security
-->
<!--- /!\ Questions should be asked on [Gitter](https://gitter.im/reactor/reactor-netty) or [StackOverflow](https://stackoverflow.com/questions/tagged/reactor-netty). -->

## Expected Behavior
<!--- Tell us what you think should happen. -->

No error

## Actual Behavior
<!--- Tell us what happens instead of the expected behavior. -->

<img width="1436" alt="no-keepalive" src="https://user-images.githubusercontent.com/5242590/87287510-c7d3c980-c52c-11ea-9f33-a47587cdd27e.png">


throws error :

```
2020-07-13 17:04:05.871 [reactor-http-nio-1] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0x0f31479e, L:/0:0:0:0:0:0:0:0:8888] READ: [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423]
2020-07-13 17:04:05.871 [reactor-http-nio-1] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0x0f31479e, L:/0:0:0:0:0:0:0:0:8888] READ COMPLETE
2020-07-13 17:04:05.872 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServerOperations - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] New http connection, requesting read
2020-07-13 17:04:05.872 [reactor-http-nio-4] DEBUG reactor.ipc.netty.channel.ContextHandler - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] After pipeline DefaultChannelPipeline{(reactor.left.loggingHandler = io.netty.handler.logging.LoggingHandler), (ServerContextHandler#0 = reactor.ipc.netty.channel.ServerContextHandler), (reactor.left.httpCodec = io.netty.handler.codec.http.HttpServerCodec), (reactor.left.httpServerHandler = reactor.ipc.netty.http.server.HttpServerHandler), (reactor.right.reactiveBridge = reactor.ipc.netty.channel.ChannelOperationsHandler)}
2020-07-13 17:04:05.872 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] REGISTERED
2020-07-13 17:04:05.872 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] ACTIVE
2020-07-13 17:04:05.873 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] READ: 124B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 47 45 54 20 2f 64 65 6d 6f 2f 74 72 61 6e 73 63 |GET /demo/transc|
|00000010| 20 48 54 54 50 2f 31 2e 31 0d 0a 43 6f 6e 6e 65 | HTTP/1.1..Conne|
|00000020| 63 74 69 6f 6e 3a 20 63 6c 6f 73 65 0d 0a 48 6f |ction: close..Ho|
|00000030| 73 74 3a 20 6c 6f 63 61 6c 68 6f 73 74 3a 38 38 |st: localhost:88|
|00000040| 38 38 0d 0a 55 73 65 72 2d 41 67 65 6e 74 3a 20 |88..User-Agent: |
|00000050| 41 70 61 63 68 65 2d 48 74 74 70 43 6c 69 65 6e |Apache-HttpClien|
|00000060| 74 2f 34 2e 35 2e 37 20 28 4a 61 76 61 2f 31 2e |t/4.5.7 (Java/1.|
|00000070| 38 2e 30 5f 31 34 34 29 0d 0a 0d 0a             |8.0_144)....    |
+--------+-------------------------------------------------+----------------+
2020-07-13 17:04:05.874 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServerOperations - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] Increasing pending responses, now 1
2020-07-13 17:04:05.875 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] READ COMPLETE
2020-07-13 17:04:05.875 [reactor-http-nio-4] DEBUG reactor.ipc.netty.channel.ChannelOperations - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] [HttpServer] Handler is being applied: org.springframework.http.server.reactive.ReactorHttpHandlerAdapter@79b11647
2020-07-13 17:04:05.875 [reactor-http-nio-4] DEBUG org.springframework.web.reactive.DispatcherHandler - Processing GET request for [http://localhost:8888/demo/transc]
2020-07-13 17:04:05.875 [reactor-http-nio-4] DEBUG org.springframework.boot.actuate.endpoint.web.reactive.WebFluxEndpointHandlerMapping - Looking up handler method for path /demo/transc
2020-07-13 17:04:05.877 [reactor-http-nio-4] DEBUG org.springframework.boot.actuate.endpoint.web.reactive.WebFluxEndpointHandlerMapping - Did not find handler method for [/demo/transc]
2020-07-13 17:04:05.877 [reactor-http-nio-4] DEBUG org.springframework.boot.actuate.endpoint.web.reactive.ControllerEndpointHandlerMapping - Looking up handler method for path /demo/transc
2020-07-13 17:04:05.878 [reactor-http-nio-4] DEBUG org.springframework.boot.actuate.endpoint.web.reactive.ControllerEndpointHandlerMapping - Did not find handler method for [/demo/transc]
2020-07-13 17:04:05.878 [reactor-http-nio-4] DEBUG org.springframework.web.reactive.result.method.annotation.RequestMappingHandlerMapping - Looking up handler method for path /demo/transc
2020-07-13 17:04:05.879 [reactor-http-nio-4] DEBUG org.springframework.web.reactive.result.method.annotation.RequestMappingHandlerMapping - Did not find handler method for [/demo/transc]
2020-07-13 17:04:05.881 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.route.RouteDefinitionRouteLocator - RouteDefinition CompositeDiscoveryClient_DEMO applying {pattern=/demo/**} to Path
2020-07-13 17:04:05.882 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'defaultValidator'
2020-07-13 17:04:05.882 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.route.RouteDefinitionRouteLocator - RouteDefinition CompositeDiscoveryClient_DEMO applying filter {regexp=/demo/(?<remaining>.*), replacement=/${remaining}} to RewritePath
2020-07-13 17:04:05.884 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.route.RouteDefinitionRouteLocator - RouteDefinition matched: CompositeDiscoveryClient_DEMO
2020-07-13 17:04:05.885 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.route.RouteDefinitionRouteLocator - RouteDefinition CompositeDiscoveryClient_SPDB-GATEWAY applying {pattern=/spdb-gateway/**} to Path
2020-07-13 17:04:05.887 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'defaultValidator'
2020-07-13 17:04:05.888 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.route.RouteDefinitionRouteLocator - RouteDefinition CompositeDiscoveryClient_SPDB-GATEWAY applying filter {regexp=/spdb-gateway/(?<remaining>.*), replacement=/${remaining}} to RewritePath
2020-07-13 17:04:05.890 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.route.RouteDefinitionRouteLocator - RouteDefinition matched: CompositeDiscoveryClient_SPDB-GATEWAY
2020-07-13 17:04:05.890 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'defaultValidator'
2020-07-13 17:04:05.891 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.handler.RoutePredicateHandlerMapping - Route matched: CompositeDiscoveryClient_DEMO
2020-07-13 17:04:05.892 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.handler.RoutePredicateHandlerMapping - Mapping [Exchange: GET http://localhost:8888/demo/transc] to Route{id='CompositeDiscoveryClient_DEMO', uri=lb://DEMO, order=0, predicate=org.springframework.cloud.gateway.support.ServerWebExchangeUtils$$Lambda$670/415647262@182b4d47, gatewayFilters=[OrderedGatewayFilter{delegate=org.springframework.cloud.gateway.filter.factory.RewritePathGatewayFilterFactory$$Lambda$672/862992864@51ad4e5c, order=1}]}
2020-07-13 17:04:05.892 [reactor-http-nio-4] DEBUG org.springframework.cloud.gateway.handler.FilteringWebHandler - Sorted gatewayFilterFactories: [OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=com.csii.spdb.filter.RequestLoggingBeginFilter@698d6d30}, order=-2147483648}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.AdaptCachedBodyGlobalFilter@10e56da9}, order=-2147482648}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.NettyWriteResponseFilter@19b9f903}, order=-1}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.ForwardPathFilter@748904e8}, order=0}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.GatewayMetricsFilter@3b3056a6}, order=0}, OrderedGatewayFilter{delegate=org.springframework.cloud.gateway.filter.factory.RewritePathGatewayFilterFactory$$Lambda$672/862992864@51ad4e5c, order=1}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.RouteToRequestUrlFilter@28cb86b2}, order=10000}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.LoadBalancerClientFilter@51ed2f68}, order=10100}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=com.csii.spdb.filter.RequestRouteLoggingFilter@3407aa4f}, order=10200}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=com.csii.spdb.filter.PreserveHostHeaderGatewayFilter@13aed42b}, order=2147483547}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.WebsocketRoutingFilter@4fbaa7f5}, order=2147483646}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=com.csii.spdb.filter.RequestLoggingEndFilter@78065fcd}, order=2147483647}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.NettyRoutingFilter@538b3c88}, order=2147483647}, OrderedGatewayFilter{delegate=GatewayFilterAdapter{delegate=org.springframework.cloud.gateway.filter.ForwardRoutingFilter@22ae905f}, order=2147483647}]
2020-07-13 17:04:05.893 [reactor-http-nio-4] INFO  com.csii.spdb.filter.RequestLoggingBeginFilter - @@@@@@@@@@ begin: GET http://localhost:8888/demo/transc
2020-07-13 17:04:05.894 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'ribbonLoadBalancer'
2020-07-13 17:04:05.895 [reactor-http-nio-4] DEBUG com.netflix.loadbalancer.ZoneAwareLoadBalancer - Zone aware logic disabled or there is only one zone
2020-07-13 17:04:05.896 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'ribbonClientConfig'
2020-07-13 17:04:05.897 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'serverIntrospector'
2020-07-13 17:04:05.898 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'serverIntrospector'
2020-07-13 17:04:05.899 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'ribbonLoadBalancerContext'
2020-07-13 17:04:05.899 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'ribbonClientConfig'
2020-07-13 17:04:05.900 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'serverIntrospector'
2020-07-13 17:04:05.901 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'forwardedHeadersFilter'
2020-07-13 17:04:05.902 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'removeHopByHopHeadersFilter'
2020-07-13 17:04:05.902 [reactor-http-nio-4] DEBUG org.springframework.beans.factory.support.DefaultListableBeanFactory - Returning cached instance of singleton bean 'xForwardedHeadersFilter'
2020-07-13 17:04:05.905 [reactor-http-nio-4] DEBUG reactor.ipc.netty.channel.PooledClientContextHandler - Acquiring existing channel from pool: DefaultPromise@27340354(incomplete) SimpleChannelPool{activeConnections=0, inactiveConnections=1}
2020-07-13 17:04:05.906 [reactor-http-nio-4] DEBUG reactor.ipc.netty.resources.DefaultPoolResources - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] Channel acquired, now 1 active connections and 0 inactive connections
2020-07-13 17:04:05.906 [reactor-http-nio-3] DEBUG reactor.ipc.netty.channel.PooledClientContextHandler - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] Acquired active channel
2020-07-13 17:04:05.906 [reactor-http-nio-3] DEBUG reactor.ipc.netty.channel.ChannelOperations - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] [HttpClient] Handler is being applied: HttpClientHandler{startURI=http://172.20.10.3:9999/transc, method=GET, handler=org.springframework.cloud.gateway.filter.NettyRoutingFilter$$Lambda$696/2073111055@2a690c77}
2020-07-13 17:04:05.906 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] USER_EVENT: reactor.ipc.netty.NettyPipeline$SendOptionsChangeEvent@14fad545
2020-07-13 17:04:05.907 [reactor-http-nio-3] DEBUG reactor.ipc.netty.channel.ChannelOperationsHandler - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] New sending options
2020-07-13 17:04:05.907 [reactor-http-nio-3] DEBUG reactor.ipc.netty.channel.ChannelOperationsHandler - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] Writing object DefaultHttpRequest(decodeResult: success, version: HTTP/1.1)
GET /transc HTTP/1.1
User-Agent: Apache-HttpClient/4.5.7 (Java/1.8.0_144)
Forwarded: proto=http;host="localhost:8888";for="127.0.0.1:52423"
X-Forwarded-For: 127.0.0.1
X-Forwarded-Proto: http
X-Forwarded-Port: 8888
X-Forwarded-Host: localhost:8888
Host: localhost:8888
2020-07-13 17:04:05.907 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] WRITE: 278B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 47 45 54 20 2f 74 72 61 6e 73 63 20 48 54 54 50 |GET /transc HTTP|
|00000010| 2f 31 2e 31 0d 0a 55 73 65 72 2d 41 67 65 6e 74 |/1.1..User-Agent|
|00000020| 3a 20 41 70 61 63 68 65 2d 48 74 74 70 43 6c 69 |: Apache-HttpCli|
|00000030| 65 6e 74 2f 34 2e 35 2e 37 20 28 4a 61 76 61 2f |ent/4.5.7 (Java/|
|00000040| 31 2e 38 2e 30 5f 31 34 34 29 0d 0a 46 6f 72 77 |1.8.0_144)..Forw|
|00000050| 61 72 64 65 64 3a 20 70 72 6f 74 6f 3d 68 74 74 |arded: proto=htt|
|00000060| 70 3b 68 6f 73 74 3d 22 6c 6f 63 61 6c 68 6f 73 |p;host="localhos|
|00000070| 74 3a 38 38 38 38 22 3b 66 6f 72 3d 22 31 32 37 |t:8888";for="127|
|00000080| 2e 30 2e 30 2e 31 3a 35 32 34 32 33 22 0d 0a 58 |.0.0.1:52423"..X|
|00000090| 2d 46 6f 72 77 61 72 64 65 64 2d 46 6f 72 3a 20 |-Forwarded-For: |
|000000a0| 31 32 37 2e 30 2e 30 2e 31 0d 0a 58 2d 46 6f 72 |127.0.0.1..X-For|
|000000b0| 77 61 72 64 65 64 2d 50 72 6f 74 6f 3a 20 68 74 |warded-Proto: ht|
|000000c0| 74 70 0d 0a 58 2d 46 6f 72 77 61 72 64 65 64 2d |tp..X-Forwarded-|
|000000d0| 50 6f 72 74 3a 20 38 38 38 38 0d 0a 58 2d 46 6f |Port: 8888..X-Fo|
|000000e0| 72 77 61 72 64 65 64 2d 48 6f 73 74 3a 20 6c 6f |rwarded-Host: lo|
|000000f0| 63 61 6c 68 6f 73 74 3a 38 38 38 38 0d 0a 48 6f |calhost:8888..Ho|
|00000100| 73 74 3a 20 6c 6f 63 61 6c 68 6f 73 74 3a 38 38 |st: localhost:88|
|00000110| 38 38 0d 0a 0d 0a                               |88....          |
+--------+-------------------------------------------------+----------------+
2020-07-13 17:04:05.908 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] FLUSH
2020-07-13 17:04:05.908 [reactor-http-nio-3] DEBUG reactor.ipc.netty.channel.ChannelOperationsHandler - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] Writing object 
2020-07-13 17:04:05.908 [reactor-http-nio-4] DEBUG reactor.ipc.netty.channel.FluxReceive - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] Subscribing inbound receiver [pending: 0, cancelled:false, inboundDone: true]
2020-07-13 17:04:05.910 [reactor-http-nio-3] DEBUG reactor.ipc.netty.channel.ChannelOperationsHandler - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] Writing object EmptyLastHttpContent
2020-07-13 17:04:05.910 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] WRITE: 0B
2020-07-13 17:04:05.910 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] FLUSH
2020-07-13 17:04:07.305 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] READ COMPLETE
2020-07-13 17:04:07.305 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 - R:/127.0.0.1:52423] EXCEPTION: java.io.IOException: Connection reset by peer
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:1.8.0_144]
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:1.8.0_144]
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223) ~[?:1.8.0_144]
	at sun.nio.ch.IOUtil.read(IOUtil.java:192) ~[?:1.8.0_144]
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380) ~[?:1.8.0_144]
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288) ~[netty-buffer-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1108) ~[netty-buffer-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:345) ~[netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:647) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:582) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:461) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884) [netty-common-4.1.25.Final.jar:4.1.25.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
2020-07-13 17:04:07.315 [reactor-http-nio-4] ERROR reactor.core.publisher.Operators - Operator called default onErrorDropped
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:1.8.0_144]
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:1.8.0_144]
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223) ~[?:1.8.0_144]
	at sun.nio.ch.IOUtil.read(IOUtil.java:192) ~[?:1.8.0_144]
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380) ~[?:1.8.0_144]
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288) ~[netty-buffer-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1108) ~[netty-buffer-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:345) ~[netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:647) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:582) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:461) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884) [netty-common-4.1.25.Final.jar:4.1.25.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
2020-07-13 17:04:07.318 [reactor-http-nio-4] DEBUG io.netty.channel.AbstractChannelHandlerContext - An exception reactor.core.Exceptions$BubblingException: java.io.IOException: Connection reset by peer
	at reactor.core.Exceptions.bubble(Exceptions.java:154)
	at reactor.core.publisher.Operators.onErrorDropped(Operators.java:263)
	at reactor.ipc.netty.channel.FluxReceive.onInboundError(FluxReceive.java:351)
	at reactor.ipc.netty.channel.ChannelOperations.onInboundError(ChannelOperations.java:429)
	at reactor.ipc.netty.channel.ChannelOperationsHandler.exceptionCaught(ChannelOperationsHandler.java:182)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireExceptionCaught(CombinedChannelDuplexHandler.java:426)
	at io.netty.channel.ChannelHandlerAdapter.exceptionCaught(ChannelHandlerAdapter.java:87)
	at io.netty.channel.CombinedChannelDuplexHandler$1.fireExceptionCaught(CombinedChannelDuplexHandler.java:147)
	at io.netty.channel.ChannelInboundHandlerAdapter.exceptionCaught(ChannelInboundHandlerAdapter.java:131)
	at io.netty.channel.CombinedChannelDuplexHandler.exceptionCaught(CombinedChannelDuplexHandler.java:233)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.handler.logging.LoggingHandler.exceptionCaught(LoggingHandler.java:175)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.AbstractChannelHandlerContext.fireExceptionCaught(AbstractChannelHandlerContext.java:256)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.exceptionCaught(DefaultChannelPipeline.java:1401)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:285)
	at io.netty.channel.AbstractChannelHandlerContext.invokeExceptionCaught(AbstractChannelHandlerContext.java:264)
	at io.netty.channel.DefaultChannelPipeline.fireExceptionCaught(DefaultChannelPipeline.java:953)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.handleReadException(AbstractNioByteChannel.java:125)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:174)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:647)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:582)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:461)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:192)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1108)
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:345)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148)
	... 6 more
was thrown by a user handler's exceptionCaught() method while handling the following exception:
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method) ~[?:1.8.0_144]
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39) ~[?:1.8.0_144]
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223) ~[?:1.8.0_144]
	at sun.nio.ch.IOUtil.read(IOUtil.java:192) ~[?:1.8.0_144]
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380) ~[?:1.8.0_144]
	at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288) ~[netty-buffer-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1108) ~[netty-buffer-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:345) ~[netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:148) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:647) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:582) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:499) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:461) [netty-transport-4.1.25.Final.jar:4.1.25.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884) [netty-common-4.1.25.Final.jar:4.1.25.Final]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
2020-07-13 17:04:07.319 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 ! R:/127.0.0.1:52423] INACTIVE
2020-07-13 17:04:07.319 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 ! R:/127.0.0.1:52423] USER_EVENT: [Handler Terminated]
2020-07-13 17:04:07.319 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 - R:/172.20.10.3:9999] CLOSE
2020-07-13 17:04:07.319 [reactor-http-nio-4] DEBUG reactor.ipc.netty.channel.ChannelOperationsHandler - [id: 0xd2b08945, L:/127.0.0.1:8888 ! R:/127.0.0.1:52423] Disposing context reactor.ipc.netty.channel.ServerContextHandler@342cb073
2020-07-13 17:04:07.319 [reactor-http-nio-4] DEBUG reactor.ipc.netty.http.server.HttpServer - [id: 0xd2b08945, L:/127.0.0.1:8888 ! R:/127.0.0.1:52423] UNREGISTERED
2020-07-13 17:04:07.320 [reactor-http-nio-3] DEBUG reactor.ipc.netty.resources.DefaultPoolResources - [id: 0x89f02c8a, L:/172.20.10.3:52421 ! R:/172.20.10.3:9999] Channel closed, now 1 active connections and -1 inactive connections
2020-07-13 17:04:07.320 [reactor-http-nio-3] DEBUG reactor.ipc.netty.resources.DefaultPoolResources - [id: 0x89f02c8a, L:/172.20.10.3:52421 ! R:/172.20.10.3:9999] Channel released, now 0 active connections and 0 inactive connections
2020-07-13 17:04:07.320 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 ! R:/172.20.10.3:9999] INACTIVE
2020-07-13 17:04:07.320 [reactor-http-nio-3] DEBUG reactor.ipc.netty.http.client.HttpClient - [id: 0x89f02c8a, L:/172.20.10.3:52421 ! R:/172.20.10.3:9999] UNREGISTERED

```
## Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to
reproduce this bug, eg. a unit test. Include code to reproduce, if relevant.
https://stackoverflow.com/help/minimal-reproducible-example -->

1. Run Eureka
2. Run Microservice (Spring boot)
3. Run Spring Cloud Gateway
4. Run Jemeter use no-keeyalive  or curl -H "Connection: Close" localhost:8888/demo/transc  to Haproxy (timeout server 500ms)

## Possible Solution
<!--- Not obligatory, but you can suggest a fix/reason for the bug. -->

I noticed that Jmeter would have no problem using Keepalive.So I keep track of the keepalive-related code。


```
# reactor.ipc.netty.http.server.HttpServerOperations.java

protected void onInboundNext(ChannelHandlerContext ctx, Object msg) {
		if (msg instanceof HttpContent) {
			if (msg != LastHttpContent.EMPTY_LAST_CONTENT) {
				super.onInboundNext(ctx, msg);
			}
			if (msg instanceof LastHttpContent) {
				onInboundComplete();
				if (!isKeepAlive()) {
					channel().config()
					         .setAutoRead(true); //detect close
				}
			}
		}
		else {
			super.onInboundNext(ctx, msg);
		}
	}
```

https://github.com/reactor/reactor-netty/commit/be41032653b7994cfa3a0a1d255fe4e503ebb4cf

The difference between KeepAlive and No KeepAlive is  **channel().config().setAutoRead(true)**。If I set it to false or delete it, no error will be reported。


I think RST triggers the event and then Netty Read triggers **java.io.IOException: Connection reset by peer** . Is that ok detect close。

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in. -->
<!--- Especially, always include the version(s) of Reactor library/libraries you used! -->

* Spring Boot version: 2.0.3.RELEASE
* Spring Cloud Version:  Finchley.SR1
* Reactor version(s) used:  0.7.11.RELEASE
* Other relevant libraries versions (eg. `netty`, ...): netty : 4.1.25.Final 
* JVM version (`javar -version`): 1.8.0_144 
* OS and version (eg `uname -a`): macOS 10.12.6 、Windows 7、CentOS 7.6

--- a/src/main/java/reactor/netty/channel/FluxReceive.java
+++ b/src/main/java/reactor/netty/channel/FluxReceive.java
@@ -385,8 +385,13 @@ final class FluxReceive extends Flux<Object> implements Subscription, Disposable
 	final void onInboundError(Throwable err) {
 		if (isCancelled() || inboundDone) {
 			if (log.isDebugEnabled()) {
-				log.warn(format(channel, "An exception has been observed post termination"), err);
-			} else if (log.isWarnEnabled()) {
+				if (AbortedException.isConnectionReset(err)) {
+					log.debug(format(channel, "Connection reset has been observed post termination"), err);
+				}
+				else {
+					log.warn(format(channel, "An exception has been observed post termination"), err);
+				}
+			} else if (log.isWarnEnabled() && !AbortedException.isConnectionReset(err)) {
 				log.warn(format(channel, "An exception has been observed post termination, use DEBUG level to see the full stack: {}"), err.toString());
 			}
 			return;

